crates/solidity/testing/sourcify/src/sourcify.rs

```diff
@@ -0,0 +1,330 @@
+}
+
+impl ArchiveDescriptor {
+    fn new(path_str: &str) -> Result<ArchiveDescriptor> {
+        // File path should come in this format:
+        // /sourcify-repository-2025-03-24T03-00-26/full_match.1.00.tar.gz
+        //                                                     - --
+        //                                                     | | shard prefix
+        //                                                     | chain ID
+        let mut parts = path_str.split('.');
+        let name_prefix = parts.next().unwrap();
+        let match_type = if name_prefix.ends_with("full_match") {
+            MatchType::Full
+        } else if name_prefix.ends_with("partial_match") {
+            MatchType::Partial
+        } else {
+            bail!("Invalid match type in archive path: {}", path_str);
+        };
+
+        let chain_id_part = parts.next().ok_or(Error::msg("Failed to get chain ID"))?;
+        let chain_id: u64 = chain_id_part.parse()?;
+
+        let shard_prefix_part = parts
+            .next()
+            .ok_or(Error::msg("Failed to get shard prefix"))?;
+        let shard_prefix = u16::from_str_radix(shard_prefix_part, 16)?;
+
+        Ok(ArchiveDescriptor {
+            path: path_str.into(),
+            shard_prefix,
+            chain_id: ChainId(chain_id),
+            match_type,
+        })
+    }
+
+    fn matches_chain_and_shard(&self, chain: Chain, options: &ShardingOptions) -> bool {
+        if self.match_type == MatchType::Partial && options.exclude_partial_matches {
+            return false;
+        }
+
+        if self.chain_id != chain.id() {
+            return false;
+        }
+
+        if !options.get_id_range().contains(&self.shard_prefix) {
+            return false;
+        }
+
+        true
+    }
+
+    /// Get a path that should be used as the target when unpacking the archive
+    /// represented by this `ArchiveDescriptor`.
+    fn archive_dir(&self) -> PathBuf {
+        PathBuf::from(format!(
+            "target/sourcify_{chain_id}/{shard_prefix:02x}",
+            chain_id = self.chain_id,
+            shard_prefix = self.shard_prefix,
+        ))
+    }
+
+    /// Get the path inside `self.archive_dir()` that contains all of the contracts.
+    /// This path is defined by the archives fetched from Sourcify, and should be updated
+    /// in case Sourcify ever changes its repository format.
+    fn contracts_dir(&self) -> PathBuf {
+        self.archive_dir().join(format!(
+            "repository/{match_type}/{chain_id}",
+            match_type = self.match_type.dir_name(),
+            chain_id = self.chain_id,
+        ))
+    }
+}
+
+#[derive(Clone)]
+pub struct ContractArchive {
+    /// Path to the directory inside this archive which contains all of the contracts.
+    /// Iterate over the entries at this path to read the contracts.
+    contracts_path: PathBuf,
+}
+
+impl ContractArchive {
+    pub fn new(desc: &ArchiveDescriptor) -> Result<ContractArchive> {
```

Comment:
I suggest renaming this function to `fetch` to make it clear that it's performing the I/O operation.